src/nupic/encoders/RandomDistributedScalarEncoder.cpp

```diff
@@ -0,0 +1,149 @@
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2019, David McDougall
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * Implementation of the RandomDistributedScalarEncoder
+ */
+
+#include <nupic/encoders/RandomDistributedScalarEncoder.hpp>
+#include <nupic/utils/MurmurHash3.hpp>
+#include <algorithm> // fill
+
+using namespace std;
+
+namespace nupic {
+namespace encoders {
+
+
+RandomDistributedScalarEncoder::RandomDistributedScalarEncoder(
+                                              const RDSE_Parameters &parameters)
+  { initialize( parameters ); }
+
+void RandomDistributedScalarEncoder::initialize( const RDSE_Parameters &parameters)
+{
+  // Check parameters
+  NTA_CHECK( parameters.size > 0u );
+
+  UInt num_active_args = 0;
+  if( parameters.activeBits > 0u)   { num_active_args++; }
+  if( parameters.sparsity   > 0.0f) { num_active_args++; }
+  NTA_CHECK( num_active_args != 0u )
+      << "Missing argument, need one of: 'activeBits' or 'sparsity'.";
+  NTA_CHECK( num_active_args == 1u )
+      << "Too many arguments, choose only one of: 'activeBits' or 'sparsity'.";
+
+  UInt num_resolution_args = 0;
+  if( parameters.radius     > 0.0f) { num_resolution_args++; }
+  if( parameters.resolution > 0.0f) { num_resolution_args++; }
+  NTA_CHECK( num_resolution_args != 0u )
+      << "Missing argument, need one of: 'radius', 'resolution'.";
+  NTA_CHECK( num_resolution_args == 1u )
+      << "Too many arguments, choose only one of: 'radius', 'resolution'.";
+
+  args_ = parameters;
+  // Finish filling in all of parameters.
+
+  // Determine number of activeBits.
+  if( args_.sparsity > 0.0f ) {
+    NTA_CHECK( args_.sparsity >= 0.0f );
+    NTA_CHECK( args_.sparsity <= 1.0f );
+    args_.activeBits = (UInt) round( args_.size * args_.sparsity );
+    NTA_CHECK( args_.activeBits > 0u );
+  }
+  // Determine sparsity. Always calculate this even if it was given, to correct for rounding error.
+  args_.sparsity = (Real) args_.activeBits / args_.size;
+
+  // Determine resolution.
+  if( args_.radius > 0.0f ) {
+    args_.resolution = args_.radius / args_.activeBits;
+  }
+  // Determine radius.
+  else if( args_.resolution > 0.0f ) {
+    args_.radius = args_.activeBits * args_.resolution;
+  }
+
+  // Initialize parent class.
+  BaseEncoder<Real64>::initialize({ args_.size });
```

Comment:
call parent initialize first (atop), then do the local stuff. It's safer in case parent/local messes with the same values.